Add user locale/region setting

tcitworld opened the pull request:
This PR allows having a dedicated user locale/region setting. See https://github.com/nextcloud/server/issues/1781

Needs :
- [x] Where to put the list of locales (obviously this locales.json is going to be moved)
- [x] More tests (please point out where)
- [x] UI fixes on personal settings page
- [x] Provide a system wide setting for locale
- [x] Add a check on OCS endpoint for valid locale format
- [x] Pass locale to MomentJS
- [x] Don't reload the page, just pass the locale to momentJS.

tcitworld committed:
Allow user to choose locale

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld committed:
Allow apps to specify locale for localisation

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld committed:
Add tests

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

nickvergessen commented on lib/private/Settings/Personal/PersonalInfo.php:
```diff
@@ -258,6 +260,50 @@ private function getLanguages(IUser $user) {
+			return $userLocaleString === $value['code'];
+		});
+
+		if (count($userLocale) > 0)
```
`!empty()`

nickvergessen commented on settings/templates/settings/personal/personal.info.php:
```diff
@@ -314,3 +315,29 @@
 </form>
 <?php } ?>
+
+<form id="locale" class="section">
```
Move above the closing bracket? You don't need the dropdown when you can't save it.

nickvergessen commented on settings/templates/settings/personal/personal.info.php:
```diff
@@ -314,3 +315,29 @@
+			</option>
+		<?php endforeach;?>
+	</select>
+</form>
```
New line

nickvergessen commented on lib/private/L10N/Factory.php:
```diff
@@ -111,7 +112,7 @@ public function get($app, $lang = null) {
 		if (!isset($this->instances[$lang][$app])) {
 			$this->instances[$lang][$app] = new L10N(
-				$this, $app, $lang,
+				$this, $app, $lang, $locale,
```
locale is not filled with the users config?!

nickvergessen commented:
> Migrate the core apps to use this

I disagree. This should happen automatically. It should automatically pick up the user's config and use that localle if none was given explicit, just like with the language. The cases where we set the language explicit are a different story of course, but there shouldn't be a lot.

tcitworld committed:
Fix review & tests

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

georgehrke commented:
>  Migrate the core apps to use this

Ideally core apps would just use momentJS and we set moment.locale($locale) in core.js

tcitworld committed:
handle config.php values

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

tcitworld committed:
add findLocale() to public interface

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld committed:
OC.getLocale() now returns Locale and no longer Language
Added OC.getLanguage() to get Language
<html lang=''> still gets language, though according to IETF BCP47 locale should be good.

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld commented:
> Ideally core apps would just use momentJS and we set moment.locale($locale) in core.js

Done.

georgehrke commented:
There is too much space between Language and Locale: 

<img width="465" alt="nextcloud chromium today at 3 39 49 pm" src="https://user-images.githubusercontent.com/1250540/28270604-623f8c4a-6b06-11e7-888d-1f66239f4c4b.png">

Additionally I would like to show some examples for the selected locale.

georgehrke commented:
AFAIK you can simply update the locale with `moment.locale(newLocale)`. There is no need to do a full reload.

georgehrke commented:
How did you acquire the data for `lib/private/L10N/locales.json`?

Is there an automated way to update it?

MorrisJobke commented:
@tcitworld Could you please resolve the conflict? Thanks 👍

tcitworld commented:
> There is too much space between Language and Locale:

Yeah, don't really know why.

> Additionally I would like to show some examples for the selected locale.

Meaning ?

> How did you acquire the data for lib/private/L10N/locales.json?

Originally stolen from https://gist.github.com/jacobbubu/1836273, but I've found newer versions at https://gist.github.com/jasef/337431c43c3addb2cbd5eb215b376179 and https://gist.github.com/ndbroadbent/588fefab8e0f1b459fcec8181b41b39c

I guess it could be retrieved from elsewhere (for instance Debian has a languages code list at `cat /usr/share/i18n/SUPPORTED`), but I didn't found any perfect up-to-date list.

>  Could you please resolve the conflict? Thanks

Fixed.

georgehrke commented:
> Meaning ?

I thought about something like this: 
<img width="730" alt="dedicated locale-region option issue 1781 nextcloud-server safari today at 2 07 48 pm" src="https://user-images.githubusercontent.com/1250540/28824565-d8320406-76c2-11e7-8e54-de7f3d4574ae.png">

doodhout commented:
Any progress on this?

Currently biding my time using en_GB as a language+locale setting, but having the option to choose both would be nice.

georgehrke commented:
@tcitworld What's the current state? :) 
Do you need any help?

georgehrke commented:
> @tcitworld What's the current state? :)
Do you need any help?

> How did you acquire the data for `lib/private/L10N/locales.json`?
Is there an automated way to update it? 

ping :)

georgehrke committed:
fix position of locale input

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke committed:
fix OC.getLocale()

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke committed:
show locale example and apply new locale right away

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

nickvergessen commented on lib/private/Settings/Personal/PersonalInfo.php:
```diff
@@ -258,6 +260,41 @@ private function getLanguages(IUser $user) {
+			'activelocaleLang' => $userLocaleString,
+			'activelocale' => $userLocale,
+			'locales' => $localeCodes,
+			'localesforlanguage' => $localesForLanguage,
```
Want to use correct CamelCasing?

nickvergessen commented on lib/private/L10N/L10N.php:
```diff
@@ -76,6 +82,15 @@ public function getLanguageCode() {
+	 * @return string locale
+	 */
+	public function getLocaleCode() {
+		return $this->locale;
```
missing the fallback for null which you do in  `l()` ? Should we do this here and use the method then below?

georgehrke replied:
Handled up here: https://github.com/nextcloud/server/pull/5623/files#diff-0adb6edb83bb2d21c924c900685ed501R128

georgehrke commented:
Looking at moment.js, if offers way fewer locales than the locale.json.
(`moment.locales()`)

Unfortunately moment.js doesn't offer readable names for each locale.
Hence I would abandon our current locale.json und generate our own based on the locales actually supported by moment.js

tcitworld commented:
In that case we could also try to translate the locale names in their own languages, like the language field does.

georgehrke commented:
We should be able to parse it from the moment-with-locales.js.
Every local contains this header: https://github.com/moment/moment/blob/develop/locale/de.js#L2

Unfortunately there is no localized version of the name.

MorrisJobke commented:
Sorry - we need to move this to 14 :/

skjnldsv commented:
It could be great to have a summary to what choosing this locale will really change, like the date, a file size example, the am/pm... etc etc :)

biva commented:
Hello, any news on this PR? Any chance to change the first day of the week soon? Thank you for your work!

biva commented:
Hello, any chance to see this in NC14? Thanks again for this work!

georgehrke commented:
There are only a couple of missing things here:
- also set local in `js.js`
- tests
- decide where list of locales should go
- rebase

> decide where list of locales should go:

Right now they are in `lib/private/L10N/locales.json`. 
@MorrisJobke Any opinion on the location of that file?

MorrisJobke commented:
> Right now they are in lib/private/L10N/locales.json.
> @MorrisJobke Any opinion on the location of that file?

we could also put it in the resources folder where also the mimetype mapping and aliases are.

georgehrke commented:
Rebased, I will move the locales.json, add moment to js.js and look into the tests.

@skjnldsv Could you help out with the design / layouting in the personal settings?
it looks a bit odd:
![bda39ed7-6d71-4018-982d-cbaa8194cb12](https://user-images.githubusercontent.com/1250540/41974122-4b2f2c60-7a17-11e8-8945-8a383bb1de9e.jpg)
![fa480bc0-cd82-409e-8b5a-2282ca5e32d3](https://user-images.githubusercontent.com/1250540/41974123-4b4d76f2-7a17-11e8-8ad9-5ea0bb00172a.jpg)

skjnldsv commented:
> @skjnldsv Could you help out with the design / layouting in the personal settings?

Of course :)

georgehrke committed:
fix tests / mistakes made during rebase

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

MorrisJobke commented:
Could you incorporate the "enforce-language" setting into this option? Just do not show the option if the language is enforced (then only admins can change it).

georgehrke committed:
fix camelCase for localesForLanguage

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke committed:
cache available locales to minimize fs access

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke committed:
properly handle force_locale setting

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

MorrisJobke commented on lib/private/L10N/Factory.php:
```diff
@@ -207,6 +207,11 @@ public function findLanguage($app = null) {
 	 * @return null|string
 	 */
 	public function findLocale($lang = null) {
+		$forceLocale = $this->config->getSystemValue('force_locale', false);
```
Please also document it in the sample config.

georgehrke replied:
That's already the case :)
https://github.com/nextcloud/server/pull/5623/files#diff-2f769bc01ffefdf14eb04404218e7582

georgehrke committed:
fix use of findLocale

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke committed:
move locales file to /resources/

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke commented:
It's done from a coding point of view imho. Just the layouting fixes :)

MorrisJobke commented on lib/public/IL10N.php:
```diff
@@ -107,4 +108,12 @@ public function l(string $type, $data, array $options = []);
+	 * * The code (en_US, fr_CA, ...) of the locale that is used for this IL10N object
+	 *
+	 * @return string locale
+	 * @since 13.0.0
```
14 ;)

MorrisJobke commented on lib/public/L10N/IFactory.php:
```diff
@@ -44,6 +44,13 @@ public function get($app, $lang = null);
+	/**
+	 * @param string|null $lang user language as default locale
+	 * @return string locale If nothing works it returns 'en_US'
+	 * @since 13.0.0
```
14 ;)

MorrisJobke commented on lib/public/L10N/IFactory.php:
```diff
@@ -53,11 +60,33 @@ public function findLanguage($app = null);
+	 *
+	 * @param string $string
+	 * @return string Unique function name
+	 * @since 9.0.0
```
And this especially ;)

MorrisJobke reviewed (approved):
Tested and works 👍  (beside the little nitpicks the code looks good as well)

skjnldsv committed:
Design fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

georgehrke committed:
docblock version fixes

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

georgehrke commented:
Fixed the docblock since mistakes

MorrisJobke committed:
Update PHPDoc

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

danger89 commented:
Does this also introduces a way for an admin to set the DEFAULT locale/region? Let's say you create a new user.

danger89 commented:
Ps. Also update the docs? https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/language_configuration.html?highlight=language#

tcitworld commented:
@danger89 
> I see, this is not an GUI option set. I think the only way is to set defaults within the config file. Correct?

Yup, just like for the language.

skjnldsv commented:
@tcitworld can you provide a pr for the docs ? :)

tcitworld commented:
@skjnldsv https://github.com/nextcloud/documentation/pull/1051
